### PR TITLE
Min realizations fix

### PR DESCRIPTION
**Pull request by flikka**
Have removed the get / set _min_realizations public functions in C and Python. Replaced with have_enough_realizations.

Changed behavior when the config MIN_REALIZATIONS is not set (is defaulted to zero), or explicitly set to zero. 

Now, zero is a no-op for this, and one require all simulations to succeed in stead.

### CODE REVIEW

File: libenkf/src/analysis_config.c [hunk 1]
```diff
@@ -125,16 +125,16 @@ ANALYSIS_SELECT  ModuleName
...
       else
         return false;
   } else {
-    /* No value has been set in the config; just compare the input with zero. */
-    if (realisations > 0)
+    /* The min_realizations is set to zero, so rather check that the number of realizations is equal to the given ensemble_size */
+    if (realisations >= ensemble_size)
       return true;
     else
       return false;
```
Comment (pgdr) [inline libenkf/src/analysis_config.c]: Perhaps simply `return realisations >= ensemble_size`?

Comment (flikka) [inline libenkf/src/analysis_config.c]: Indeedably!

File: libenkf/src/enkf_main.c [hunk 2]
```diff
@@ -1286,9 +1286,8 @@ bool enkf_main_UPDATE(enkf_main_type * enkf_main , const int_vector_type * step_
     int_vector_free( ens_active_list );
     return true;
   } else {
-    fprintf(stderr,"** ERROR ** There are %d active realisations left, which is less than the minimum specified (%d) - stopping assimilation.\n" ,
-            active_ens_size ,
-            analysis_config_get_min_realisations(analysis_config));
+    fprintf(stderr,"** ERROR ** There are %d active realisations left, which is less than the minimum specified - stopping assimilation.\n" ,
+            active_ens_size );
```
Comment (pgdr) [inline libenkf/src/enkf_main.c]: did we lose the min_specified information in this context, or:  Why is it removed from the error message?

Comment (flikka) [inline libenkf/src/enkf_main.c]: Yeah - it is gone... (as you can see from the changes to analysis_config.{c|h})

File: libenkf/src/enkf_main.c [hunk 3]
```diff
@@ -1534,12 +1530,11 @@ void enkf_main_submit_jobs( enkf_main_type * enkf_main ,
 
 
 /**
-  If all simulations have completed successfully the function will
-  return true, otherwise it will return false.
+  The function will return number of non-failing jobs.
```
Comment (pgdr) [inline libenkf/src/enkf_main.c]: Is #non-failing-jobs = #succeeding-jobs ?

Comment (flikka) [inline libenkf/src/enkf_main.c]: I guess it should be, but the actual thing being counted is "Total - (explicitly failed)", so I called it non-failing.

File: python/python/ert/enkf/analysis_config.py [hunk 4]
```diff
@@ -123,41 +116,43 @@ def setGlobalStdScaling(self, std_scaling):
...
+    def haveEnoughRealisations(self, realizations, ensemble_size):
+        return AnalysisConfig.cNamespace().have_enough_realisations(self, realizations, ensemble_size)
+
 
 cwrapper = CWrapper(ENKF_LIB)
 cwrapper.registerType("analysis_config", AnalysisConfig)
 cwrapper.registerType("analysis_config_obj", AnalysisConfig.createPythonObject)
 cwrapper.registerType("analysis_config_ref", AnalysisConfig.createCReference)
 
-AnalysisConfig.cNamespace().alloc                  = cwrapper.prototype("c_void_p analysis_config_alloc()")
```
Comment (jepebe) [inline python/python/ert/enkf/analysis_config.py]: I suggest you convert the class to new style c wrapping

File: python/python/ert_gui/shell/simulations.py [hunk 5]
```diff
@@ -36,10 +36,11 @@ def ensembleExperiment(self, line):
 
         print("Ensemble Experiment started at: %s" % datetime.now().isoformat(sep=" "))
 
-        success = simulation_runner.runEnsembleExperiment()
+        num_successful_realizations = simulation_runner.runEnsembleExperiment()
 
+        success = self.ert().analysisConfig().haveEnoughRealisations(num_successful_realizations, self.ert().getEnsembleSize())
         if not success:
-            print("Error: Simulations failed!")
+            print("Error: Number of successful realizations is too low, you can change this by altering the MIN_REALIZATIONS element!")
```
Comment (jepebe) [inline python/python/ert_gui/shell/simulations.py]: If the number is `0` will I need to set `MIN_REALIZATIONS` to `0` as well?

Comment (flikka) [inline python/python/ert_gui/shell/simulations.py]: I think you mean that the error message is unclear. I agree! Will change

File: libenkf/applications/ert_tui/main.c [hunk 6]
```diff
@@ -78,7 +78,7 @@ void enkf_usage() {
   printf(" **                            E R T                                **\n");
   printf(" **                                                                 **\n");
   printf(" **-----------------------------------------------------------------**\n");
-  printf(" ** You have sucessfully started the ert program developed at       **\n");
+  printf(" ** You have successfully started the ert program developed at       **\n");
```
Comment (jepebe) [inline libenkf/applications/ert_tui/main.c]: The stars are not aligned!

Comment (flikka) [inline libenkf/applications/ert_tui/main.c]: ... well, two days after next full moon, they'll be perfectly aligned!

**Comment by flikka** [comment]
Will squash once reviewed...

File: libenkf/applications/ert_tui/main.c (hunk 6 above)
Comment (pgdr) [inline libenkf/applications/ert_tui/main.c]: Do not try and align the stars, that's impossible.

Instead, only try to realize the truth … there are no stars.

File: libenkf/src/analysis_config.c [hunk 7]
```diff
@@ -125,19 +125,13 @@ ANALYSIS_SELECT  ModuleName
...
-  } else {
-    /* No value has been set in the config; just compare the input with zero. */
-    if (realisations > 0)
-      return true;
-    else
-      return false;
+    return realisations >= config->min_realisations;
+  } 
+  else {
+    return realisations >= ensemble_size;
```
Comment (jepebe) [inline libenkf/src/analysis_config.c]: Can `realisations` ever be larger than `ensemble_size`?

File: python/python/ert/enkf/enkf_simulation_runner.py [hunk 8]
```diff
@@ -46,4 +46,4 @@ def runWorkflows(self , runtime):
 
 EnkfSimulationRunner.cNamespace().run_smoother      = cwrapper.prototype("void enkf_main_run_smoother(enkf_simulation_runner, char*, bool)")
 EnkfSimulationRunner.cNamespace().create_run_path   = cwrapper.prototype("bool enkf_main_create_run_path(enkf_simulation_runner, bool_vector, int)")
-EnkfSimulationRunner.cNamespace().run_simple_step   = cwrapper.prototype("bool enkf_main_run_simple_step(enkf_simulation_runner, bool_vector, enkf_init_mode_enum, int)")
+EnkfSimulationRunner.cNamespace().run_simple_step   = cwrapper.prototype("int enkf_main_run_simple_step(enkf_simulation_runner, bool_vector, enkf_init_mode_enum, int)")
```
Comment (jepebe) [inline python/python/ert/enkf/enkf_simulation_runner.py]: Also a candidate for updated wrapping but not required.

**Comment by jepebe** [comment]
Have you considered the simulation code found in `ertrpcserver.py`?

File: libenkf/src/analysis_config.c (hunk 7 above)
Comment (flikka) [inline libenkf/src/analysis_config.c]: Probably not.

File: python/python/ert/enkf/enkf_simulation_runner.py (hunk 8 above)
Comment (flikka) [inline python/python/ert/enkf/enkf_simulation_runner.py]: I'll do it!

**Comment by flikka** [comment]
@jepebe: Will look at ertrpcserver.py

File: libenkf/src/analysis_config.c (hunk 7 above)
Comment (joakim-hove) [inline libenkf/src/analysis_config.c]: I agree with @jepebe here - using:
```
return (realisations == ensemble_size);
```
would be clearer.

File: libenkf/tests/enkf_analysis_config.c [hunk 9]
```diff
@@ -46,20 +46,22 @@ void test_create() {
...
+void test_min_realizations(const char * num_realizations_str, const char * min_realizations_str, int min_realizations_expected_needed) {
+  test_work_area_type * work_area = test_work_area_alloc("test_min_realizations_string");
 
   {
     FILE * config_file_stream = util_mkdir_fopen("config_file", "w");
     test_assert_not_NULL(config_file_stream);
-    fprintf(config_file_stream, num_realizations_str);
-    fprintf(config_file_stream, min_realizations_str);
+
+    fputs(num_realizations_str, config_file_stream);
```
Comment (joakim-hove) [inline libenkf/tests/enkf_analysis_config.c]: Why using `fputs( )`instead of `fprintf( )`?

**Pull request merged**